### PR TITLE
ENH: `escape` html argument in Styler.format

### LINKED ISSUE #40388: Add raw_html parameter to Styler.format, enabling richer output
`Styler.format` could allow a parameter `raw_html` (bool, False for current behaviour), enabling richer output

For example, `df.style.format(lambda s: f"<a href='https://en.wikipedia.org/wiki/{html.escape(s)}'>{html.escape(s)}</a>", raw_html=True)` could be used to turn a column of values into a column of links to English Wikipedia articles, when exported with `to_html`.

To achieve this, we would need to always pass raw HTML to the jina templates, changing how formatting is currently done in `Styler.to_html`.

Formats performed with `raw_html=True` would have no effect on `Styler.to_excel`.

### PR DESCRIPTION
**Pull request by attack68**
- [x] closes #40388 

Adds an HTML `escape` parameter to `Styler.format`, which acts to escape the data value **before** it is passed to a formatter.

For example, `Styler.format('<a href="a.com/{0}">{0}</a>', escape=True)` will not escape the HTML as part of the formatting function, it will only escape the value `{0}` going into the formatter.

### CODE REVIEW

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -113,6 +114,10 @@ class Styler:
 
         .. versionadded:: 1.2.0
 
+    escape : bool, default False
+        Replace the characters ``&``, ``<``, ``>``, ``'``, and ``"`` in cell display
+        strings with HTML-safe sequences.
+
```
Comment (jreback) [inline pandas/io/formats/style.py:120]: versionadded 1.3

**Review by jreback — CHANGES_REQUESTED** [review]
sure, minor comment. ping on green.

**Comment by jreback** [comment]
cc @jnothman if comments.

**Comment by jreback** [comment]
also checks are failing

**Review by jnothman — APPROVED** [review]
LGTM

**Comment by attack68** [comment]
> also checks are failing

when fixing these I noticed that the `na_rep` was not working with `escape` since missing values were escaped before being replaced.

so I refactored to fix this and added a test for it.

@jreback ping is green.

File: pandas/io/formats/style.py [hunk 2]
```diff
@@ -2171,9 +2206,12 @@ def _maybe_wrap_formatter(
         raise TypeError(f"'formatter' expected str or callable, got {type(formatter)}")
 
     if na_rep is None:
-        return formatter_func
+        return (lambda x: formatter_func(escape_func(x))) if escape else formatter_func
     else:
-        return lambda x: na_rep if pd.isna(x) else formatter_func(x)
+        if escape:
+            return lambda x: na_rep if pd.isna(x) else formatter_func(escape_func(x))
```
Comment (attack68) [inline pandas/io/formats/style.py:2212]: This might be a bug escaping floats, since escaping converts to string. Will review and come back

Comment (attack68) [reply pandas/io/formats/style.py]: OK this is fixed now too. 3rd test added.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/formats/style.py [hunk 3]
```diff
@@ -2170,10 +2205,17 @@ def _maybe_wrap_formatter(
     else:
         raise TypeError(f"'formatter' expected str or callable, got {type(formatter)}")
 
+    def _str_escape(x):
+        """only use escape_func on str, else return input"""
+        return escape_func(x) if isinstance(x, str) else x
+
     if na_rep is None:
```
[the remark this answers is not in the document]
Comment (attack68) [reply pandas/io/formats/style.py:2217]: hows this?

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/io/formats/style.py:2217]: good

**Comment by attack68** [comment]
this should be good to go in now, I believe.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @attack68

**Pull request merged**